### PR TITLE
Flatten state dictionary for proper weight loading

### PR DESCRIPTION
**Pull request by degenfabian**
# Description
While I was working on #859, I noticed that with the new weight conversions, the accuracy of the models is severely decreased. The reason for that is that the way the new weight conversions work, they create a nested state dictionary, where the different weights for the layers of the model are stored as list elements of the dictionary attribute "blocks" (or e.g. "encoder" and "decoder" for the T5 architecture). PyTorch's function ```load_state_dict()```can't handle the state dictionaries when they are input with this nested structure. Therefore, the weights within the layers of the model get assigned zero tensors or NaN values, or other unwanted behavior. This PR fixes this by flattening out the dictionary before it is passed to the ```load_state_dict()``` function.

There is no specific issue attached to this PR.

## Illustration
Here you can see the differences in logits for the different layers before and after the change, for the BERT model. I also tested it on the T5 model and got similar results.

Results before the change:
```
Analyzing resid_post_embed similarity:
Equality: True
Maximum difference: 1.9073486e-06
Mean difference: 3.1006973e-08

Analyzing resid_post_first_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_second_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_third_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_fourth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_fifth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_sixth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_seventh_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_eighth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_ninth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_tenth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_eleventh_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing resid_post_twelvth_layer similarity:
Equality: False
Maximum difference: nan
Mean difference: nan

Analyzing final_logits similarity:
Equality: False
Maximum difference: nan
Mean difference: nan
```

Results after the change:
```
Analyzing resid_post_embed similarity:
Equality: True
Maximum difference: 1.9073486e-06
Mean difference: 3.1006973e-08

Analyzing resid_post_first_layer similarity:
Equality: True
Maximum difference: 2.3841858e-06
Mean difference: 2.4683123e-07

Analyzing resid_post_second_layer similarity:
Equality: True
Maximum difference: 1.335144e-05
Mean difference: 3.2591592e-07

Analyzing resid_post_third_layer similarity:
Equality: True
Maximum difference: 7.1525574e-06
Mean difference: 3.6711398e-07

Analyzing resid_post_fourth_layer similarity:
Equality: True
Maximum difference: 2.9802322e-06
Mean difference: 4.205786e-07

Analyzing resid_post_fifth_layer similarity:
Equality: True
Maximum difference: 4.7683716e-06
Mean difference: 4.779429e-07

Analyzing resid_post_sixth_layer similarity:
Equality: True
Maximum difference: 3.8146973e-06
Mean difference: 5.107744e-07

Analyzing resid_post_seventh_layer similarity:
Equality: True
Maximum difference: 5.722046e-06
Mean difference: 5.4364614e-07

Analyzing resid_post_eighth_layer similarity:
Equality: True
Maximum difference: 1.04904175e-05
Mean difference: 5.7116586e-07

Analyzing resid_post_ninth_layer similarity:
Equality: True
Maximum difference: 1.5258789e-05
Mean difference: 5.9130906e-07

Analyzing resid_post_tenth_layer similarity:
Equality: True
Maximum difference: 9.536743e-06
Mean difference: 5.773467e-07

Analyzing resid_post_eleventh_layer similarity:
Equality: True
Maximum difference: 9.536743e-06
Mean difference: 5.562378e-07

Analyzing resid_post_twelvth_layer similarity:
Equality: True
Maximum difference: 6.198883e-06
Mean difference: 4.4152608e-07

Analyzing final_logits similarity:
Equality: True
Maximum difference: 4.2915344e-05
Mean difference: 5.0457793e-06
```

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Comment by bryce13950** [comment]
@degenfabian I went ahead and merged this to get things moving with the weight conversion, but before the whole refactor is wrapped up, the functionality you added to loading_from_pretrained should really be put into the actual conversion. What comes out of the conversion should be the fully workable state_dict, so if you want to move it, that would be a relatively simple task to complete.
